### ISSUE TITLE
False positive for unstable_name_collisions lint

### ISSUE THREAD

**Issue by coolreader18** [rust-lang/rust#83834]
Cargo.toml:
```
[package]
name = "playground"
version = "0.1.0"
edition = "2018"

[dependencies]
lexical-core = "0.7"
```
lib.rs:
```rust
fn link_lexical_core() {
    let _ = lexical_core::get_inf_string();
}

fn huh() {
    let _x = 1u32 >> 32.wrapping_sub(1usize) as u32;
}
```
on `cargo c`:
```
warning: a method with this name may be added to the standard library in the future
 --> src/lib.rs:6:25
  |
6 |     let _x = 1u32 >> 32.wrapping_sub(1usize) as u32;
  |                         ^^^^^^^^^^^^
  |
  = note: `#[warn(unstable_name_collisions)]` on by default
  = warning: once this method is added to the standard library, the ambiguity may cause an error or change in behavior!
  = note: for more information, see issue #48919 <https://github.com/rust-lang/rust/issues/48919>
  = help: call with fully qualified syntax `lexical_core::Integer::wrapping_sub(...)` to keep using the current method

error[E0689]: can't call method `wrapping_sub` on ambiguous numeric type `{integer}`
 --> src/lib.rs:6:25
  |
6 |     let _x = 1u32 >> 32.wrapping_sub(1usize) as u32;
  |                         ^^^^^^^^^^^^
  |
help: you must specify a concrete type for this numeric value, like `i32`
  |
6 |     let _x = 1u32 >> 32_i32.wrapping_sub(1usize) as u32;
  |                      ^^^^^^

error: aborting due to previous error; 1 warning emitted
```
Fixing the original error by adding a suffix to the integer literal gets rid of the unstable_name_collisions warning as well. This doesn't give the unstable_name_collisions warning:
```rust
fn link_lexical_core() {
    let _ = lexical_core::get_inf_string();
}

fn huh() {
    let _x = 1u32 >> 32usize.wrapping_sub(1usize) as u32;
}
```
Maybe rustc is noticing that `Integer` is implemented for exactly the same set of types that a bare integer literal could be? And it assumes that since that method wasn't fully resolved, that it's an unstable method, and doesn't check that it's already stable?

**Comment by eggyal** [comment]
What version of rustc is this?  1.51 and nightly both produce [E0689](https://doc.rust-lang.org/stable/error-index.html#E0689) instead, at least on [a pared down example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4cfe901b53568a53740e128d1a50bd80).

**Comment by eggyal** [comment]
Okay, what's happening here...

In `rustc_typeck::check::method::probe::consider_candidates`, there are two `applicable_candidates`:
* `compiler_builtins::int::Int::wrapping_sub`; and
* `lexical_core::util::num::Integer::wrapping_sub`.

The first of these is unstable:

``DEBUG rustc_middle::middle::stability stability: inspecting def_id=DefId(3:285 ~ compiler_builtins[702f]::int::Int::wrapping_sub) span=src/main.rs:6:25: 6:37 (#0) of stability=Some(Stability { level: Unstable { reason: Some("this crate is being loaded from the sysroot, an unstable location; did you mean to load this crate from crates.io via `Cargo.toml` instead?"), issue: Some(27812), is_soft: false }, feature: "rustc_private" })``

And so there actually is a collision against an unstable name, and the emitted warning is correct.  Adding a suffix to the integer literal changes the builtin candidate from the unstable one above to that on the relevant primitive (which is stable).

Would emitting the path of the clashing unstable item make this diagnostic clearer?

**Comment by jonas-schievink** [comment]
None of those traits are in scope though, so I'm not sure why they would ever even be considered for method dispatch

**Comment by eggyal** [comment]
@jonas-schievink:

> None of those traits are in scope though, so I'm not sure why they would ever even be considered for method dispatch

It's [only for diagnostics](https://github.com/rust-lang/rust/blob/62a49c3bb8e0c82ee40b482f08a21a65f90b874b/compiler/rustc_typeck/src/check/method/probe.rs#L1047)...

<pre>
├─31ms DEBUG rustc_typeck::check::method::probe pick_method(self_ty={integer})
├─31ms DEBUG rustc_typeck::check::method::probe searching inherent candidates
├─31ms DEBUG rustc_typeck::check::method::probe applicable_candidates: []
├─31ms DEBUG rustc_typeck::check::method::probe searching extension candidates
├─31ms DEBUG rustc_typeck::check::method::probe applicable_candidates: []
├─31ms DEBUG rustc_typeck::check::method::probe searching unstable candidates
├─31ms DEBUG rustc_typeck::check::method::probe applicable_candidates: []
├─31ms DEBUG rustc_typeck::check::method::probe pick_method(self_ty=&{integer})
├─31ms DEBUG rustc_typeck::check::method::probe searching inherent candidates
├─31ms DEBUG rustc_typeck::check::method::probe applicable_candidates: []
├─31ms DEBUG rustc_typeck::check::method::probe searching extension candidates
├─31ms DEBUG rustc_typeck::check::method::probe applicable_candidates: []
├─31ms DEBUG rustc_typeck::check::method::probe searching unstable candidates
├─31ms DEBUG rustc_typeck::check::method::probe applicable_candidates: []
├─31ms DEBUG rustc_typeck::check::method::probe pick_method(self_ty=&mut {integer})
├─31ms DEBUG rustc_typeck::check::method::probe searching inherent candidates
├─31ms DEBUG rustc_typeck::check::method::probe applicable_candidates: []
├─32ms DEBUG rustc_typeck::check::method::probe searching extension candidates
├─32ms DEBUG rustc_typeck::check::method::probe applicable_candidates: []
├─32ms DEBUG rustc_typeck::check::method::probe searching unstable candidates
├─32ms DEBUG rustc_typeck::check::method::probe applicable_candidates: []
<strong>├─32ms DEBUG rustc_typeck::check::method::probe pick: actual search failed, assemble diagnostics
├─771ms DEBUG rustc_typeck::check::method::probe assemble_extension_candidates_for_trait</strong>
...
├─1007ms DEBUG rustc_typeck::check::method::probe assemble_extension_candidates_for_trait(trait_def_id=DefId(3:263 ~ compiler_builtins[702f]::int::Int))
... 
├─1037ms DEBUG rustc_typeck::check::method::probe assemble_extension_candidates_for_trait(trait_def_id=DefId(16:242 ~ lexical_core[1557]::util::num::Integer))
</pre>
